packages/esm-patient-orders-app/src/lab-results/useLabResultsFormSchema.tsx

```diff
@@ -0,0 +1,124 @@
+    return createSetMembersSchema(setMembers);
+  }
+
+  return createSingleConceptSchema(concept);
+};
+
+/**
+ * Creates a Zod schema for a single lab order concept.
+ * @param labOrderConcept - The lab order concept to create a schema for.
+ * @returns A Zod schema object for the single concept.
+ */
+const createSingleConceptSchema = (labOrderConcept: LabOrderConcept) => {
+  const { hiAbsolute: upperLimit, lowAbsolute: lowerLimit, datatype } = labOrderConcept;
+  if (datatype.display?.toLowerCase() === 'numeric') {
+    const zodObject = createNumericSchema(labOrderConcept, upperLimit, lowerLimit);
+
+    return z.object({
+      [labOrderConcept.uuid]: zodObject,
+    });
+  }
+
+  return z.object({
+    [labOrderConcept.uuid]: z.any(),
+  });
+};
+
+/**
+ * Creates a Zod schema for a set of lab order concepts.
+ * @param labOrderConcepts - An array of lab order concepts.
+ * @returns A Zod schema object for the set of concepts.
+ */
+const createSetMembersSchema = (labOrderConcepts: Array<LabOrderConcept>): z.ZodObject<SchemaRecord> => {
+  const schema = z.object(
+    labOrderConcepts.reduce<SchemaRecord>((acc, member) => {
+      acc[member.uuid] = createSchema(member);
+      return acc;
+    }, {}),
+  );
+
+  return schema;
+};
+
+/**
+ * Determines and returns the appropriate Zod schema for a given lab order concept.
+ * @param labOrderConcept - The lab order concept to create a schema for.
+ * @returns A Zod schema type based on the concept's data type.
+ */
+const createSchema = (labOrderConcept: LabOrderConcept): z.ZodType => {
+  const conceptDataType = labOrderConcept.datatype.display?.toLowerCase();
+  const { hiAbsolute: upperLimit, lowAbsolute: lowerLimit } = labOrderConcept;
+  if (conceptDataType === 'text') {
+    return z.string().optional();
+  }
+
+  if (conceptDataType === 'numeric') {
+    return createNumericSchema(labOrderConcept, upperLimit, lowerLimit);
+  }
+
+  // Default case
+  return z.unknown();
+};
+
+/**
+ * Creates a Zod schema for a numeric lab order concept, including validation for upper and lower limits.
+ * @param labOrderConcept - The lab order concept to create a schema for.
+ * @param upperLimit - The upper limit for the numeric value.
+ * @param lowerLimit - The lower limit for the numeric value.
+ * @returns A Zod schema type with appropriate numeric validations.
+ */
+const createNumericSchema = (
+  labOrderConcept: LabOrderConcept,
+  upperLimit: number | undefined,
+  lowerLimit: number | undefined,
+): z.ZodType => {
+  let baseSchema = z
+    .preprocess((val) => {
+      if (val === '' || val === null || val === undefined) return undefined;
+      const parsed = Number(val);
+      return isNaN(parsed) ? undefined : parsed;
+    }, z.number().optional())
+    .refine((val) => val === undefined || !isNaN(val), {
+      message: `${labOrderConcept.display} must be a valid number`,
+    });
+
+  if (lowerLimit === null && upperLimit === null) {
+    return baseSchema;
+  }
```

Comment:
Shouldn't the check here be for `undefined` rather than `null`? The typings imply that these variables can never be null.

Reply:
I'm under the impression that what we accept here should be driven by what the backend actually returns, i.e., if we're always setting these values to `null` (as in the tests), then the types here should be `number | null` and I guess the code works as-is. If the backend doesn't return values for those, then the types should be as-written, but the code updated. The safest option would be to just explicitly support `number | null | undefined` and then be sure to check for undefined.

Reply:
Thanks, @ibacher